pkg/operator/mon/mon.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	err = c.startPods(clientset, clusterInfo, mons)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start mon pods. %+v", err)
+	}
+
+	return clusterInfo, nil
+}
+
+// Retrieve the ceph cluster info if it already exists.
+// If a new cluster create new keys.
+func (c *Cluster) initClusterInfo() (*mon.ClusterInfo, error) {
+	return mon.CreateClusterInfo(c.factory, "")
+}
+
+func (c *Cluster) startPods(clientset *kubernetes.Clientset, clusterInfo *mon.ClusterInfo, mons []*MonConfig) error {
+	// schedule the mons on different nodes if we have enough nodes to be unique
+	antiAffinity, err := c.getAntiAffinity(clientset)
+	if err != nil {
+		return fmt.Errorf("failed to get antiaffinity. %+v", err)
+	}
+
+	running, pending, err := c.pollPods(clientset, clusterInfo.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get mon pods. %+v", err)
+	}
+	logger.Infof("%d running, %d pending pods", len(running), len(pending))
+
+	if len(running) == c.Size {
+		logger.Infof("pods are already running")
+		return nil
+	}
+
+	clusterInfo.Monitors = map[string]*mon.CephMonitorConfig{}
+	for _, m := range running {
+		clusterInfo.Monitors[m.Name] = mon.ToCephMon(m.Name, m.Status.PodIP)
+	}
+
+	started := 0
+	alreadyRunning := 0
+	for _, m := range mons {
+		monPod := c.makeMonPod(m, clusterInfo, antiAffinity)
+		logger.Debugf("Starting pod: %+v", monPod)
+		_, err := clientset.Pods(c.Namespace).Create(monPod)
+		if err != nil {
+			if !k8sutil.IsKubernetesResourceAlreadyExistError(err) {
+				return fmt.Errorf("failed to create mon pod %s. %+v", c.Namespace, err)
+			}
+			alreadyRunning++
+			logger.Infof("mon pod %s already exists", monPod.Name)
+		} else {
+			started++
+		}
+
+		podIP, err := c.waitForPodToStart(clientset, monPod)
+		if err != nil {
+			return fmt.Errorf("failed to start pod %s. %+v", monPod.Name, err)
+		}
+		clusterInfo.Monitors[m.Name] = mon.ToCephMon(m.Name, podIP)
+	}
+
+	logger.Infof("started %d/%d mons (%d already running)", (started + alreadyRunning), c.Size, alreadyRunning)
+	return nil
+}
+
+func (c *Cluster) waitForPodToStart(clientset *kubernetes.Clientset, pod *v1.Pod) (string, error) {
+
+	// Poll the status of the pods to see if they are ready
+	// FIX: Get status instead of just waiting
+	for i := 0; i < 15; i++ {
+		// wait and try again
+		delay := 8
+		logger.Infof("waiting %ds for pod %s to start. status=%v", delay, pod.Name, pod.Status.Phase)
+		<-time.After(time.Duration(delay) * time.Second)
+
+		pod, err := clientset.Core().Pods(c.Namespace).Get(pod.Name)
+		if err != nil {
+			return "", fmt.Errorf("failed to get mon pod %s. %+v", pod.Name, err)
+		}
+
+		if pod.Status.Phase == v1.PodRunning {
+			logger.Infof("pod %s started", pod.Name)
+			return pod.Status.PodIP, nil
+		}
+	}
+
+	return "", fmt.Errorf("timed out waiting for pod %s to start", pod.Name)
+}
+
+func (c *Cluster) getAntiAffinity(clientset *kubernetes.Clientset) (bool, error) {
```

Comment:
a comment here on what anti affinity is (or at least what it means in the context this function should be called from) would be nice